src/geff/spatial_graph/io.py

```diff
@@ -168,7 +170,7 @@ def construct_sg(
     node_props: dict[str, PropDictNpArray],
     edge_props: dict[str, PropDictNpArray],
     position_attr: str = "position",
-) -> sg.SpatialGraph:
+) -> sg.SpatialGraph | sg.SpatialDiGraph:
```

Comment:
the remaining type error you're getting is indicating that `read_sg` needs to be updated to reflect the new return type of `construct_sg` (which is correct):

```python
-> tuple[sg.SpatialGraph | sg.SpatialDiGraph, GeffMetadata]
```

(because sg.SpatialDiGraph is not a subclass of sg.SpatialGraph here)